### PR TITLE
Orca: Add data loader examples to NCF with merged features

### PR DESCRIPTION
**Pull request by zpeng1898**
## Description

Add data loader examples to NCF. 

### 1. Why the change?

To show how to use `pandas.dataframe.merge` API to add features to datasets.

### 2. User API changes

No change.

### CODE REVIEW

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 1]
```diff
@@ -23,69 +23,106 @@
...
-                while (u, j) in self.train_mat:
-                    j = np.random.randint(self.num_item)
-                self.features_ng.append([u, j])
-
-        labels_ps = [1 for _ in range(len(self.features_ps))]
-        labels_ng = [0 for _ in range(len(self.features_ng))]
-
-        self.features_fill = self.features_ps + self.features_ng
-        self.labels_fill = labels_ps + labels_ng
+    def __init__(self, data):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:30]: we'd better keep the original arguments in init?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 2]
```diff
@@ -23,69 +23,106 @@
...
+        self.data["label"] = labels
+            
+    def merge_features(self, users, movies, total_cols):
+        self.data = users.merge(self.data, on='user')
+        self.data = self.data.merge(movies, on='item')
+        self.data = self.data.loc[:, total_cols]
+
+
+def load_dataset(dataset_dir, cal_sparse_feats_input_dims=True,
+                 num_ng=4, merge_features=True):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:69]: remove the merge_features options, we will always merge them

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 3]
```diff
@@ -23,69 +23,98 @@
...
-    # load ratings as a dok matrix
-    train_mat = sp.dok_matrix((user_num, item_num), dtype=np.int64)
-    for x in data_X.values.tolist():
-        train_mat[x[0], x[1]] = 1
-
-    return data_X, user_num, item_num, train_mat
+    movies = pd.read_csv(
+        os.path.join(dataset_dir, 'movies.dat'),
+        sep="::", header=None, names=['item', 'category'],
+        usecols=[0, 2], dtype={0: np.int64, 1: np.object})
```
[the remark this answers is not in the document]
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:93]: the embedding layer needs the type of inputs to be long(int64) .

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 4]
```diff
@@ -23,69 +23,98 @@
...
+    feature_cols = ['user', 'item',
+                    'gender', 'occupation', 'zipcode', 'category',  # sparse features
+                    'age']  # dense features
+    label_cols = ["label"]
+
+    users = pd.read_csv(
+        os.path.join(dataset_dir, 'users.dat'),
+        sep="::", header=None, names=['user', 'gender', 'age', 'occupation', 'zipcode'],
+        usecols=[0, 1, 2, 3, 4],
+        dtype={0: np.int64, 1: np.object, 2: np.int64, 3: np.int64, 4: np.object})
```
[the remark this answers is not in the document]
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:85]: There is no np.string in numpy datatype.

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 2 above)
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:69]: The option is used for model_creator(), because calculating user_num and item_num in it doesn't need to merge features.

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hkvision) [reply python/orca/tutorial/NCF/pytorch_dataset.py:89]: I think there is np.str

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 2 above)
Comment (hkvision) [reply python/orca/tutorial/NCF/pytorch_dataset.py:73]: I think you can extract a function like `process_user_item` that handles the user/item, which can be used for the model?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 5]
```diff
@@ -23,69 +23,112 @@
...
-        label = float(labels[idx])
-        return user, item, label
-
-
-def load_dataset(dataset_dir):
-    data_X = pd.read_csv(
+        return tuple(self.data[idx])
+
+
+def load_dataset(dataset_dir, cal_sparse_feats_input_dims=True,
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:72]: cal_sparse_feats_input_dims this is not a good argument, better parse to a function

File: python/orca/tutorial/NCF/pytorch_train_dataloader.py [hunk 6]
```diff
@@ -34,41 +37,54 @@
...
+    dataset, user_num, item_num, \
+        sparse_feats_input_dims = load_dataset(config['dataset_dir'],
+                                               cal_sparse_feats_input_dims=False,
+                                               num_ng=config['num_ng'],
+                                               merge_features=True)
 
     # train test split
-    train_data, _ = train_test_split(data_X, test_size=0.2, random_state=100)
+    dataset.data = list(map(lambda row: list(row[1:]), dataset.data.itertuples()))
+    train_dataset, _ = train_test_split(dataset, test_size=0.2, random_state=100)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_train_dataloader.py:48]: can train_test_split be performed on the PyTorch Dataset?

File: python/orca/tutorial/NCF/pytorch_train_dataloader.py [hunk 7]
```diff
@@ -34,41 +37,54 @@
...
+    dataset, user_num, item_num, \
+        sparse_feats_input_dims = load_dataset(config['dataset_dir'],
+                                               cal_sparse_feats_input_dims=False,
+                                               num_ng=config['num_ng'],
+                                               merge_features=True)
 
     # train test split
-    _, test_data = train_test_split(data_X, test_size=0.2, random_state=100)
+    dataset.data = list(map(lambda row: list(row[1:]), dataset.data.itertuples()))
+    _, test_dataset = train_test_split(dataset, test_size=0.2, random_state=100)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_train_dataloader.py:64]: Can we directly split the raw ratings?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 8]
```diff
@@ -22,70 +22,143 @@
...
+    for i in sparse_features:
+        df = users if i in users.columns else items
+        df[i], _ = pd.Series(df[i]).factorize()
+    sparse_features.append('occupation')  # occupation is already indexed.
+
+    # scale dense features
+    for i in dense_features:
+        scaler = MinMaxScaler()
+        df = users if i in users.columns else items
+        values = df[i].values.reshape(-1, 1)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:107]: why need reshape here?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 9]
```diff
@@ -22,70 +22,143 @@
...
+        values = scaler.fit_transform(values)
+        values = [torch.tensor(v, dtype=torch.float32) for v in values]
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:109]: it is not reasonable to convert to torch tensor in dataset?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 10]
```diff
@@ -22,70 +22,143 @@
...
+    dataset = NCFData(ratings, item_num, train_mat, num_ng)
+    dataset.ng_sample()
+
+    # train test split
+    train_dataset, test_dataset = dataset.train_test_split()
 
-    return data_X, user_num, item_num, train_mat
+    # merge features
+    train_dataset.merge_features(users, items, total_cols)
+    test_dataset.merge_features(users, items, total_cols)
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:160]: first merge and then split?

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 8 above)
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:107]: Because MinMaxScaler needs the input to be 2-dim tensor, not 1-dim.

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 9 above)
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:109]: If the type of the input of each feature is list, then the args of forward() in NCF model will be list automatically. However, the layer in the model just supports the input of type tensor not list. 
e.g. [0.345] will be converted to [tensor([0.345])] automatically. But tensor([0.345]) will not be converted to [tensor([0.345])].

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 11]
```diff
@@ -22,70 +22,143 @@
...
+        self.data['label'] = labels_fill
+
+    def train_test_split(self):
+        train_dataset, test_dataset = train_test_split(self.data, test_size=0.2, random_state=100)
+        return NCFData(train_dataset), NCFData(test_dataset)
+
+    def merge_features(self, users, items, total_cols):
+        self.data = users.merge(self.data, on='user')
+        self.data = self.data.merge(items, on='item')
+        self.data = self.data.loc[:, total_cols]
```
[the remark this answers is not in the document]
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:71]: The order of df columns after the merge() is not guaranteed to be the same as the original one.

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 10 above)
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:160]: image
merge only after ratings is split.

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 9 above)
Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:109]: image

Comment (zpeng1898) [reply python/orca/tutorial/NCF/pytorch_dataset.py:109]: image

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 12]
```diff
@@ -22,70 +22,142 @@
...
+                features_ng.append([u, j])
+        labels_ps = [1.0 for _ in range(len(features_ps))]
+        labels_ng = [0.0 for _ in range(len(features_ng))]
+        features_fill = features_ps + features_ng
+        labels_fill = labels_ps + labels_ng
+        self.data = pd.DataFrame(features_fill,
+                                 columns=["user", "item"], dtype=np.int32)
+        self.data['label'] = labels_fill
+
+    def train_test_split(self):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:64]: add test_size=0.2 as an argument?

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 13]
```diff
@@ -22,70 +22,142 @@
...
+        train_dataset, test_dataset = train_test_split(self.data, test_size=0.2, random_state=100)
+        return NCFData(train_dataset), NCFData(test_dataset)
+
+    def merge_features(self, users, items, total_cols):
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:68]: default total_cols=None
and if not None you call loc?

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hkvision) [reply python/orca/tutorial/NCF/pytorch_dataset.py:71]: Add a comment `# To make the order of data columns as expected.`

File: python/orca/tutorial/NCF/pytorch_train_dataloader.py [hunk 14]
```diff
@@ -82,19 +78,23 @@ def optimizer_creator(model, config):
 # Step 4: Distributed training with Orca PyTorch Estimator
 dataset_dir = "./ml-1m"
 backend = "ray"  # "ray" or "spark"
+callbacks = [TensorBoardCallback(log_dir="runs",
+                                 freq="epoch")]
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_train_dataloader.py:82]: put it into one line?

File: python/orca/tutorial/NCF/pytorch_dataset.py (hunk 8 above)
Comment (hkvision) [reply python/orca/tutorial/NCF/pytorch_dataset.py]: Makes sense.

File: python/orca/tutorial/NCF/pytorch_train_dataloader.py [hunk 15]
```diff
@@ -82,6 +78,7 @@ def optimizer_creator(model, config):
 # Step 4: Distributed training with Orca PyTorch Estimator
 dataset_dir = "./ml-1m"
 backend = "ray"  # "ray" or "spark"
+callbacks = [TensorBoardCallback(log_dir="runs", freq="epoch")]
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_train_dataloader.py:81]: change to batch with freq=1000

File: python/orca/tutorial/NCF/pytorch_dataset.py [hunk 16]
```diff
@@ -22,70 +22,148 @@
 import pandas as pd
 import scipy.sparse as sp
 
+import torch
```
Comment (hkvision) [inline python/orca/tutorial/NCF/pytorch_dataset.py:25]: is this needed?
